### PR TITLE
Provide Kubeflow PyTorchJob support for MultiKueue

### PR DESCRIPTION
**Pull request by mszadkow**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
The PR introduces a new MultiKueue adapter to handle PyTorchJob (Kubeflow).
We want to extend MultiKueue capabilities to satisfy the needs of early adopters.

#### Which issue(s) this PR fixes:

Relates #2552

#### Does this PR introduce a user-facing change?

```release-note
MultiKueue: Support for the Kubeflow PyTorchJob
```

### CODE REVIEW

File: test/e2e/multikueue/e2e_test.go [hunk 1]
```diff
@@ -459,6 +460,88 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 				}, util.Timeout, util.Interval).Should(gomega.Succeed())
 			})
 		})
+
+		ginkgo.It("Should run a kubeflow PyTorchJob on worker if admitted", func() {
+			pyTorchJob := testingpytorchjob.MakePyTorchJob("pytorchjob1", managerNs.Name).
```
Comment (trasc) [inline test/e2e/multikueue/e2e_test.go:465]: Add a comment on the worker we expect o run it and why

 (nit: maybe we should make this one run in worker 2)

**Review by trasc — COMMENTED** [review]
overall LGTM

File: test/e2e/multikueue/e2e_test.go (hunk 1 above)
Comment (mszadkow) [reply test/e2e/multikueue/e2e_test.go:465]: make sense I will update in all of the jobs

**Review by trasc — COMMENTED** [review]
/lgtm

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	remoteJob = &kftraining.PyTorchJob{
+		ObjectMeta: api.CloneObjectMetaForCreation(&localJob.ObjectMeta),
+		Spec:       *localJob.Spec.DeepCopy(),
+	}
+
+	// add the prebuilt workload
+	if remoteJob.Labels == nil {
+		remoteJob.Labels = map[string]string{}
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go:70]:
```suggestion
		remoteJob.Labels = make(map[string]string, 2)
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		return client.IgnoreNotFound(err)
+	}
+	return client.IgnoreNotFound(remoteClient.Delete(ctx, &job))
+}
+
+func (b *multikueueAdapter) KeepAdmissionCheckPending() bool {
+	return false
+}
+
+func (b *multikueueAdapter) IsJobManagedByKueue(ctx context.Context, c client.Client, key types.NamespacedName) (bool, string, error) {
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go:91]:
```suggestion
func (b *multikueueAdapter) IsJobManagedByKueue(context.Context, client.Client, types.NamespacedName) (bool, string, error) {
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+var _ jobframework.MultiKueueWatcher = (*multikueueAdapter)(nil)
+
+func (*multikueueAdapter) GetEmptyList() client.ObjectList {
+	return &kftraining.PyTorchJobList{}
+}
+
+func (*multikueueAdapter) WorkloadKeyFor(o runtime.Object) (types.NamespacedName, error) {
+	pyTorchJob, isPyTorchJob := o.(*kftraining.PyTorchJob)
+	if !isPyTorchJob {
+		return types.NamespacedName{}, errors.New("not a PYTorch Job")
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go:108]:
```suggestion
		return types.NamespacedName{}, errors.New("not a PyTorchJob")
```

File: pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	prebuiltWl, hasPrebuiltWorkload := pyTorchJob.Labels[constants.PrebuiltWorkloadLabel]
+	if !hasPrebuiltWorkload {
+		return types.NamespacedName{}, fmt.Errorf("no prebuilt workload found for PYTorch Job: %s", klog.KObj(pyTorchJob))
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorch_multikueue_adapter.go:113]:
```suggestion
		return types.NamespacedName{}, fmt.Errorf("no prebuilt workload found for PyTorchJob: %s", klog.KObj(pyTorchJob))
```

**Review by tenzen-y — COMMENTED** [review]
Just some nits.

**Comment by mszadkow** [comment]
@tenzen-y @trasc updated, ptal

**Review by tenzen-y — COMMENTED** [review]
Thanks!
/lgtm
/approve

**Pull request merged**